frappe/utils/global_search.py

```diff
@@ -419,32 +419,48 @@ def search(text, start=0, limit=20, doctype=""):
 	:param limit: number of results to return, default 20
 	:return: Array of result objects
 	"""
+	from frappe.desk.doctype.global_search_settings.global_search_settings import get_doctypes_for_global_search
+
 	results = []
-	texts = text.split('&')
+	texts = [t.strip() for t in text.split('&') if t]
+	priorities = get_doctypes_for_global_search()
+	allowed_doctypes = ",".join(["'{0}'".format(dt) for dt in priorities])
 	for text in texts:
 		mariadb_conditions = ''
 		postgres_conditions = ''
+
 		if doctype:
 			mariadb_conditions = postgres_conditions = '`doctype` = {} AND '.format(frappe.db.escape(doctype))
 
-		mariadb_conditions += 'MATCH(`content`) AGAINST ({} IN BOOLEAN MODE)'.format(frappe.db.escape('+' + text + '*'))
-		postgres_conditions += 'TO_TSVECTOR("content") @@ PLAINTO_TSQUERY({})'.format(frappe.db.escape(text))
+		mariadb_text = frappe.db.escape('+' + text + '*')
 
-		common_query = '''SELECT `doctype`, `name`, `content`
-					FROM `__global_search`
-					WHERE {conditions}
-					LIMIT {limit} OFFSET {start}'''
+		mariadb_fields = '`doctype`, `name`, `content`, MATCH (`content`) AGAINST ({} IN BOOLEAN MODE) AS rank'.format(mariadb_text)
+		postgres_fields = '`doctype`, `name`, `content`, TO_TSVECTOR("content") @@ PLAINTO_TSQUERY({}) AS rank'.format(frappe.db.escape(text))
+
+		if allowed_doctypes:
+			mariadb_conditions += '`doctype` IN ({})'.format(allowed_doctypes)
```

Comment:
IN clause might slow things down. Test the query performance before and after this change. I think a sub-query here might be faster.